torch/_functorch/aot_autograd.py

```diff
@@ -1754,12 +1772,14 @@ def functionalized_f_helper(*args):
                     # We found an input that had a (data-only) mutation.
                     # Since keep_input_mutations is set, we need to faithfully apply a copy_()
                     # so the compiler will see the input mutation in the graph.
-                    assert inpt_new is not inpt_old
-                    if meta.input_info[i].mutations_hidden_from_autograd:
-                        with torch.no_grad(), torch.autograd._unsafe_preserve_version_counter(inpt_old):
+                    if not trace_joint:
+                        assert inpt_new is not inpt_old
+                    if not trace_joint or (trace_joint and not inpt_old.requires_grad):
```

Comment:
Add an assert that `not inpt_new.requires_grad`

Reply:
I'm not sure this assert is true

Reply:
So the example I am interested in is this:

```
def f(x, y):
   x.add_(y)
```

where x does not require grad and y requires grad.

On entry to create_functionalized_fn, I expect to get args with requires_grad (False, True). When I am done, I expect f_args for x to now require grad (because I inplace mutated it with a tensor that requires grad.) I am not sure if you can somehow do it soundly, but is certainly safest to NOT persist copy_ in this case.

Now, the potential difficulty in the code below is that you are only testing `inpt_old.requires_grad`. Which I would expect to be False for x, even for the example above. So this seems a bit suspicious.

Reply:
We checked and it turned out this is wrong, but input_info[i].requires_grad has the requires grad info AFTER the function is run and works; but this needs to be documented in the InputInfo

Reply:
comments from discuss on why cloning is necessary

```
def f(x)
  x.mul_(2)
  return x

y = f(x)
torch.autograd.grad(y, x)  # identity (1)

x_new = x.clone()
y = f(x_new)
torch.autograd.grad(y, x)  # 2
```